File: loading-button-android/src/main/java/br/com/simplepass/loading_button_lib/animatedDrawables/CircularRevealAnimatedDrawable.java
package br.com.simplepass.loading_button_lib.animatedDrawables;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.animation.DecelerateInterpolator;

/**
 * Created by Leandro Ferreira on 22/09/16.
 *
 * An animation of the 'Done' status. It makes a circular reveal of a background color and the
 * than show the image passed in the constructor.
 */
public class CircularRevealAnimatedDrawable extends Drawable implements Animatable {

    private boolean mIsFilled;
    private Paint mPaint;
    private Paint mPaintImageReady;
    private View mAnimatedView;
    private float mRadius;
    private float mFinalRadius;
    private ValueAnimator mRevealInAnimation;
    private boolean isRunning;
    private float mCenterWidth;
    private float mCenterHeith;
    private Bitmap mReadyImage;
    private int mImageReadyAlpha;
    private float bitMapXOffset;
    private float bitMapYOffset;

    /**
     *
     * @param view The view that if being animated
     * @param fillColor The color of the background that will the revealed
     * @param bitmap The animage that will be shown in the end of the animation.
     */
    public CircularRevealAnimatedDrawable(View view, int fillColor, Bitmap bitmap) {
        mAnimatedView = view;
        isRunning = false;

        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setColor(fillColor);

        mPaintImageReady = new Paint();
        mPaintImageReady.setAntiAlias(true);
        mPaintImageReady.setStyle(Paint.Style.FILL);
        mPaintImageReady.setColor(Color.TRANSPARENT);

        mReadyImage = bitmap;
        mImageReadyAlpha = 0;

        mRadius = 0;
    }

    public boolean isFilled() {
        return mIsFilled;
    }

    /**
     * The method is called when bounds change
     *
     * @param bounds
     */
    @Override
    protected void onBoundsChange(Rect bounds) {
        super.onBoundsChange(bounds);

        int bitMapWidth = (int) ((bounds.right - bounds.left) * 0.6);
        int bitMapHeight = (int) ((bounds.bottom - bounds.top) * 0.6);

        bitMapXOffset =(float) (((bounds.right - bounds.left) - bitMapWidth)/2);
        bitMapYOffset =(float) (((bounds.bottom - bounds.top) - bitMapHeight)/2);

        mReadyImage = Bitmap.createScaledBitmap(mReadyImage, bitMapWidth, bitMapHeight, false);

        mFinalRadius = (bounds.right - bounds.left)/2;
        mCenterWidth = (bounds.right + bounds.left)/2;
        mCenterHeith = (bounds.bottom + bounds.top)/2;
    }

    /**
     * Setup all the animations. There are a reveal animation to show the button background
     * and a alpha animation to show the bitmap.
     */
    private void setupAnimations(){
        final ValueAnimator alphaAnimator = ValueAnimator.ofInt(0, 255);
        alphaAnimator.setDuration(80);
        alphaAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                mImageReadyAlpha = (int) animation.getAnimatedValue();
                invalidateSelf();
                mAnimatedView.invalidate();
            }
        });

        mRevealInAnimation = ValueAnimator.ofFloat(0, mFinalRadius);
        mRevealInAnimation.setInterpolator(new DecelerateInterpolator());
        mRevealInAnimation.setDuration(120);
        mRevealInAnimation.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                mRadius = (float) animation.getAnimatedValue();
                invalidateSelf();
                mAnimatedView.invalidate();
            }
        });
        mRevealInAnimation.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                super.onAnimationEnd(animation);
                mIsFilled = true;

                alphaAnimator.start();
            }
        });
    }

    /**
     * Starts the animation
     */
    @Override
    public void start() {
        if (isRunning()) {
            return;
        }

        setupAnimations();
        isRunning = true;
        mRevealInAnimation.start();
    }

    /**
     * Stops the animation
     */
    @Override
    public void stop() {
        if (!isRunning()) {
            return;
        }

        isRunning = false;
        mRevealInAnimation.cancel();
    }

    /**
     *
     * @return Return if its running or not.
     */
    @Override
    public boolean isRunning() {
        return isRunning;
    }

    @Override
    public void draw(Canvas canvas) {
        canvas.drawCircle(mCenterWidth, mCenterHeith, mRadius, mPaint);

        if(mIsFilled) {
            mPaintImageReady.setAlpha(mImageReadyAlpha);
            canvas.drawBitmap(mReadyImage, bitMapXOffset, bitMapYOffset, mPaintImageReady);
        }
    }

    @Override
    public void setAlpha(int alpha) {}

    @Override
    public void setColorFilter(ColorFilter colorFilter) {}

    @Override
    public int getOpacity() {
        return PixelFormat.OPAQUE;
    }

    public void dispose() {
        if (mRevealInAnimation != null) {
            mRevealInAnimation.end();
            mRevealInAnimation.removeAllUpdateListeners();
            mRevealInAnimation.cancel();
        }

        mRevealInAnimation = null;
    }
}

File: loading-button-android/src/main/java/br/com/simplepass/loadingbutton/animatedDrawables/CircularRevealAnimatedDrawable.kt
package br.com.simplepass.loadingbutton.animatedDrawables

import android.animation.*
import android.graphics.*
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import android.view.animation.DecelerateInterpolator
import br.com.simplepass.loadingbutton.customViews.ProgressButton
import br.com.simplepass.loadingbutton.disposeAnimator

private const val REVEAL_DURATION = 120L
private const val ALPHA_ANIMATION_DURATION = 80L

internal class CircularRevealAnimatedDrawable(
    private val progressButton: ProgressButton,
    fillColor: Int,
    image: Bitmap
) : Drawable(), Animatable {

    private var currentRadius = 0F
    private var isFilled = false
    private var imageReadyAlpha = 0

    private val finalRadius: Float by lazy { (bounds.right - bounds.left).toFloat() / 2 }
    private val centerWidth: Float by lazy { (bounds.right + bounds.left).toFloat() / 2 }
    private val centerHeight: Float by lazy { (bounds.bottom + bounds.top).toFloat() / 2 }

    private val readyImage: Bitmap by lazy {
        Bitmap.createScaledBitmap(image, bitMapWidth().toInt(), bitMapHeight().toInt(), false)
    }

    private val bitMapXOffset: Float by lazy {
        (centerWidth - bitMapWidth() / 2).toFloat()
    }

    private val bitMapYOffset: Float by lazy {
        ((centerHeight - bitMapHeight() / 2)).toFloat()
    }

    private val conclusionAnimation: AnimatorSet by lazy {
        AnimatorSet().apply {
            playSequentially(
                revealAnimator(finalRadius, DecelerateInterpolator()),
                alphaAnimator()
            )
        }
    }

    private val paint = Paint().apply {
        isAntiAlias = true
        style = Paint.Style.FILL
        color = fillColor
    }

    private val imageReadyPaint = Paint().apply {
        isAntiAlias = true
        style = Paint.Style.FILL
        color = Color.TRANSPARENT
    }

    private fun bitMapWidth(): Double = ((bounds.right - bounds.left) * 0.6)

    private fun bitMapHeight(): Double = ((bounds.bottom - bounds.top) * 0.6)

    private fun revealAnimator(radius: Float, timeInterpolator: TimeInterpolator): Animator =
        ValueAnimator.ofFloat(0F, radius).apply {
            interpolator = timeInterpolator
            duration = REVEAL_DURATION

            addUpdateListener { animation ->
                currentRadius = animation.animatedValue as Float
                progressButton.invalidate()
            }

            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator?) {
                    super.onAnimationEnd(animation)
                    isFilled = true
                }
            })
        }

    private fun alphaAnimator(): Animator =
        ValueAnimator.ofInt(0, 255).apply {
            duration = ALPHA_ANIMATION_DURATION
            addUpdateListener { animation ->
                imageReadyAlpha = animation.animatedValue as Int
                progressButton.invalidate()
            }
        }

    override fun draw(canvas: Canvas) {
        canvas.drawCircle(centerWidth, centerHeight, currentRadius, paint)

        if (isFilled) {
            imageReadyPaint.alpha = imageReadyAlpha
            canvas.drawBitmap(readyImage, bitMapXOffset, bitMapYOffset, imageReadyPaint)
        }
    }

    override fun setAlpha(alpha: Int) {}

    override fun getOpacity(): Int = PixelFormat.OPAQUE

    override fun setColorFilter(colorFilter: ColorFilter?) {}

    override fun isRunning(): Boolean = conclusionAnimation.isRunning

    override fun start() {
        conclusionAnimation.start()
    }

    override fun stop() {
        conclusionAnimation.end()
    }

    fun dispose() {
        conclusionAnimation.disposeAnimator()
    }
}
